File: flutter/lib/src/options.dart
import 'cursor_store.dart';
import 'logging.dart';
import 'protocol.dart';
import 'socket.dart';

/// Everything the SDK needs to open a connection, and the hooks a host app must supply.
class ImOptions {
  const ImOptions({
    required this.endpoint,
    required this.appId,
    required this.token,
    required this.deviceId,
    required this.userId,
    required this.cursorStore,
    this.platform = ImPlatform.unknown,
    this.channel = '/im',
    this.clientVersion,
    this.language,
    this.requestTimeout = const Duration(seconds: 15),
    this.maxAutoRepairSeq = 500,
    this.cursorSaveDebounce = const Duration(milliseconds: 1000),
    this.logger = defaultImLogger,
    this.onTokenExpired,
    this.socketFactory,
  });

  /// Base URL, e.g. `wss://im.example.com`. The channel path is appended.
  final String endpoint;

  final String appId;

  /// Short-lived user token minted by the tenant's own backend. Never an app secret.
  final String token;

  /// Stable per installation. The multi-device policy uses it to tell a reconnect from a second
  /// device, so a value that changes on every launch makes users kick themselves offline.
  final String deviceId;

  /// The signed-in user.
  ///
  /// Not sent in the handshake — the gateway reads the identity out of the token — but required,
  /// because the cursor store is scoped `(endpoint host, appId, userId)` and without the user id
  /// account switching on a shared handset hands one user the other's cursors (§5.3). Your backend
  /// already knows the value: it is the user it just minted a token for.
  ///
  /// 必填：游标作用域是 (host, appId, userId)，缺了 userId 就无法阻止同设备换账号时游标串号。
  final String userId;

  /// Where per-conversation cursors survive a process restart. **Required, with no default.**
  ///
  /// Without persistence the SDK cold-starts knowing nothing, `conn.sync` reports no gap for a
  /// conversation it was told nothing about, the client adopts the server's newest `maxSeq`, and
  /// every message that arrived while the app was closed is skipped — silently. That is a real bug
  /// this SDK shipped with, and a required argument is the only fix that cannot be forgotten.
  ///
  /// Use `ImCursorStore.file(path)` with a directory from `path_provider`, keyed by
  /// [ImCursorScope.storageKey]; or `ImCursorStore.inMemory()` if you genuinely do not want
  /// persistence and have read why that costs messages.
  ///
  /// 必填且无默认值：不持久化游标就会在冷启动时静默丢掉离线期间的全部消息。
  final ImCursorStore cursorStore;

  final ImPlatform platform;

  final String channel;

  /// Sent as the handshake's `cv`. Defaults to this package's own version, so a support ticket
  /// carries it without anyone having to ask. Set it to the *host app's* version if that is more
  /// useful to you.
  final String? clientVersion;

  /// BCP-47. The gateway remembers it, and offline push falls back to it when
  /// `push.register` carries no language of its own.
  final String? language;

  final Duration requestTimeout;

  /// Largest gap repaired message-by-message.
  ///
  /// Beyond it the SDK advances the cursor to the conversation's current `maxSeq`, commits, and
  /// raises [ImClient.conversationNeedsReload] so the application reloads that conversation from
  /// history. **Both halves matter**: skipping the cursor advance makes every later message look
  /// like a gap and re-request a range you have already declined; skipping the event leaves a hole
  /// in the UI that nothing will ever mention.
  ///
  /// Raising this above 500 does not buy a bigger single `msg.sync` — the server clamps that at
  /// 500 — it buys more iterations of the repair loop. That is fine and intended.
  final int maxAutoRepairSeq;

  /// How long ordinary [ImClient.commit] calls may be coalesced before they must reach the cursor
  /// store. Floor `Duration.zero` (write through); no ceiling.
  ///
  /// Adoption writes are never debounced regardless of this. The asymmetry is deliberate: **a lost
  /// debounced commit costs one duplicate delivery; a lost adoption costs silent permanent data
  /// loss.**
  final Duration cursorSaveDebounce;

  /// Where the SDK's own warnings go. Two of them are contractual — the in-memory cursor store and
  /// an unregistered push token — and both are things an integrator must be able to see.
  final ImLogger logger;

  /// Called when a token has expired: on a `conn.reauth` opportunity mid-session, and when the
  /// server closes with `im-kick:TokenExpired`. Return a fresh token, or null to stop.
  ///
  /// With this set, an expired token costs one `conn.reauth` frame on the socket that is already
  /// open rather than a full reconnect — which on the flaky network where tokens tend to expire is
  /// exactly what you were trying to avoid.
  final Future<String?> Function()? onTokenExpired;

  /// Supplies the transport. Defaults to `package:web_socket_channel`; tests and hosts with their
  /// own socket stack override it.
  final ImSocketFactory? socketFactory;

  ImOptions copyWith({String? token}) => ImOptions(
        endpoint: endpoint,
        appId: appId,
        token: token ?? this.token,
        deviceId: deviceId,
        userId: userId,
        cursorStore: cursorStore,
        platform: platform,
        channel: channel,
        clientVersion: clientVersion,
        language: language,
        requestTimeout: requestTimeout,
        maxAutoRepairSeq: maxAutoRepairSeq,
        cursorSaveDebounce: cursorSaveDebounce,
        logger: logger,
        onTokenExpired: onTokenExpired,
        socketFactory: socketFactory,
      );
}

/// Where the connection currently is. Drives a "reconnecting…" banner and nothing else.
enum ImConnectionState { idle, connecting, open, reconnecting, closed }

File: swift/Sources/CyaimIM/Options.swift
import Foundation

/// Everything the client needs to open a connection and keep it open.
public struct ImClientOptions: Sendable {
    /// Gateway base URL, e.g. `wss://im.example.com`. `https`/`http` are rewritten to `wss`/`ws`,
    /// because the two get mixed up constantly and the failure is a silent one.
    public var endpoint: URL

    /// Channel path appended to `endpoint`. Only change this if the tenant's gateway is mounted
    /// somewhere other than `/im`.
    public var channelPath: String

    /// Tenant identifier. Safe to ship in the app; the `AppSecret` never is.
    public var appId: String

    /// Short-lived user token minted by the tenant's backend. Never an app secret.
    public var token: String

    /// Stable per-installation identifier.
    ///
    /// The multi-device policy uses this to tell a reconnect apart from a second device. A value
    /// regenerated per launch makes users kick themselves offline, so persist it in the keychain
    /// (not `UserDefaults`, which is wiped by a reinstall while the keychain is not).
    public var deviceId: String

    /// The signed-in user. **Required.**
    ///
    /// Not sent in the handshake — the gateway reads the identity out of the token — but the third
    /// component of the cursor store's scope, `(endpoint host, appId, userId)`, and the one that
    /// stops account switching on a shared device from handing one user another's cursors
    /// (§5.3). Your backend already knows the value: it is the user it just minted a token for.
    public var userId: String

    public var platform: Platform

    /// Your app's version, surfaced in the console's session list. Worth setting: it is how you
    /// find out that a bug only happens on a build you shipped three releases ago.
    public var clientVersion: String?

    /// BCP-47 language tag. The server localises system messages and push copy with it.
    public var language: String?

    /// How long an unanswered request waits before it throws `ImError(code: .timeout)`.
    public var requestTimeout: Duration

    /// Starting heartbeat cadence, used only until the server states its own in the first reply.
    public var heartbeatInterval: Duration

    /// Reconnect delay policy. Full jitter; see `FullJitterBackoff`.
    public var backoff: FullJitterBackoff

    /// Largest gap the client repairs message by message.
    ///
    /// Past this, the conversation is too far behind to backfill: both cursors jump to the newest
    /// seq, the SDK raises ``ImSessionEvent/conversationNeedsReload(conversationId:fromSeq:toSeq:)``,
    /// and the app re-opens the conversation from history instead. Replaying fifty thousand
    /// messages into a UI helps nobody, and doing it on a phone on cellular helps less.
    ///
    /// Raising this above 500 does not buy a bigger single `msg.sync` — the server clamps that to
    /// 500 — it buys more iterations of the repair loop, which is fine and intended.
    public var maxAutoRepairSeq: Int64

    /// Where cursors live between launches. **Required, with no default.**
    ///
    /// There is no implicit default because defaulting to no persistence is exactly what produced
    /// the cold-start data-loss bug in four of the five SDKs, and defaulting to *some* persistence
    /// would mean the SDK guessing where your app is allowed to write. A compile error is the
    /// cheapest possible place to have this conversation. Pass ``ImCursorStore/inMemory()`` if you
    /// really do not want persistence; you get one warning and nothing else.
    ///
    /// 没有默认值：默认不持久化正是那个静默丢消息的 bug，而默认某种持久化等于替应用猜它能往哪写。
    public var cursorStore: ImCursorStore

    /// How long an ordinary commit may sit before the store is asked to save. Floor 0, no ceiling.
    ///
    /// A flush always happens before `conn.sync` regardless, and an *adoption* write is never
    /// debounced at all. The asymmetry is the point: a lost debounced commit costs one duplicate
    /// delivery, while a lost adoption write costs silent permanent data loss.
    public var cursorFlushInterval: Duration

    /// How many `msg.sync` pages one repair may take before it gives up and declares the gap
    /// oversized instead. Guards against a server that always reports `hasMore`.
    public var maxRepairPages: Int

    /// How many `conn.sync` pages one resume may take. At the default page size of 200 this is
    /// 20,000 conversations, which is well past any real account and still bounded.
    public var maxResumePages: Int

    /// Where the SDK's handful of warnings go. `nil` writes them to standard error.
    ///
    /// There are only a few, they are all one-per-session, and every one names the section of the
    /// contract behind it — because each warns about something that is otherwise completely silent.
    public var warningHandler: (@Sendable (String) -> Void)?

    /// Called when the server reports the token has expired, and when the gateway refuses the
    /// upgrade with 401.
    ///
    /// Return a fresh token from your backend to reconnect, or `nil` to stop trying — `nil` is the
    /// right answer when the user has been signed out of your app entirely.
    public var tokenProvider: (@Sendable () async -> String?)?

    /// Where the SDK's own runtime log lives between launches. **Optional, and the default is honest.**
    ///
    /// Unlike ``cursorStore`` this has a default, and the asymmetry is deliberate: losing cursors
    /// loses a user's messages, while losing logs loses a diagnostic. Supplying one is what makes
    /// "pull a log from that handset" answer questions about anything before the current launch — a
    /// crash, a night the app was closed, the reconnect storm at 3am. Supplying none keeps a bounded
    /// ring in memory, and the console shows a support engineer which of the two they are reading.
    ///
    /// The SDK does not pick a location for you — see `ADR-003`. On iOS that would be `Documents`
    /// or `Library/Caches`, whose backup behaviour differs, and "do chat logs reach iCloud" is a
    /// question you answer to a regulator. ``ImLogStore/file(at:maxBytes:)`` is there for when you
    /// have made that decision.
    ///
    /// SDK 不替你选写入位置：不提供也是一个完整的选择，而控制台会把这个区别显示出来。
    public var logStore: ImLogStore

    /// Transport factory. The default opens a `URLSessionWebSocketTask`; tests substitute their own.
    public var connector: any ImWebSocketConnector

    public init(
        endpoint: URL,
        appId: String,
        token: String,
        deviceId: String,
        userId: String,
        cursorStore: ImCursorStore,
        platform: Platform = .current,
        channelPath: String = "/im",
        clientVersion: String? = nil,
        language: String? = nil,
        requestTimeout: Duration = .seconds(15),
        heartbeatInterval: Duration = .seconds(30),
        backoff: FullJitterBackoff = .default,
        maxAutoRepairSeq: Int64 = 500,
        cursorFlushInterval: Duration = .milliseconds(1000),
        maxRepairPages: Int = 16,
        maxResumePages: Int = 100,
        warningHandler: (@Sendable (String) -> Void)? = nil,
        tokenProvider: (@Sendable () async -> String?)? = nil,
        logStore: ImLogStore = .inMemory(),
        connector: any ImWebSocketConnector = URLSessionWebSocketConnector()
    ) {
        self.endpoint = endpoint
        self.appId = appId
        self.token = token
        self.deviceId = deviceId
        self.userId = userId
        self.cursorStore = cursorStore
        self.logStore = logStore
        self.platform = platform
        self.channelPath = channelPath
        self.clientVersion = clientVersion
        self.language = language
        self.requestTimeout = requestTimeout
        self.heartbeatInterval = heartbeatInterval
        self.backoff = backoff
        self.maxAutoRepairSeq = maxAutoRepairSeq
        self.cursorFlushInterval = cursorFlushInterval
        self.maxRepairPages = maxRepairPages
        self.maxResumePages = maxResumePages
        self.warningHandler = warningHandler
        self.tokenProvider = tokenProvider
        self.connector = connector
    }

    /// What goes in the `cv` handshake parameter.
    ///
    /// Always carries this package's version, and the host app's too when it set one. The gateway
    /// stores it verbatim on the session, so a support ticket arrives with both halves of "which
    /// build is this" without anyone having to ask for either.
    /// Who this session's cursors belong to.
    var cursorScope: ImCursorScope {
        ImCursorScope(endpoint: endpoint, appId: appId, userId: userId)
    }

    var handshakeClientVersion: String {
        guard let clientVersion, !clientVersion.isEmpty else { return ImSdk.userAgent }
        return "\(clientVersion) \(ImSdk.userAgent)"
    }
}

/// Where the socket is in its lifecycle.
///
/// `reconnecting` is deliberately distinct from `connecting`: an app should show a banner for one
/// and a spinner for the other, and telling them apart after the fact is impossible.
public enum ConnectionState: String, Sendable, Hashable, CustomStringConvertible {
    /// Never connected, or explicitly disconnected and awaiting a foreground.
    case idle

    /// First connection attempt in flight.
    case connecting

    /// Socket is up and requests will be answered.
    case open

    /// Dropped; a jittered retry is scheduled.
    case reconnecting

    /// Closed for good — either the app asked, or the server said do not come back.
    case closed

    public var description: String { rawValue }

    /// True when a request issued right now would fail immediately.
    public var isUsable: Bool { self == .open }
}
